servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/DefaultGrpcServerBuilder.java

```diff
@@ -155,23 +186,71 @@ protected void doAppendHttpServiceFilter(final Predicate<StreamingHttpRequest> p
         httpServerBuilder.appendServiceFilter(predicate, factory);
     }
 
+    private HttpServerBuilder preBuild() {
+        if (!invokedBuild && null != defaultTimeout && GRPC_MAX_TIMEOUT.compareTo(defaultTimeout) >= 0) {
+            doAppendHttpServiceFilter(new TimeoutHttpServiceFilter(grpcDetermineTimeout()));
+        }
+        invokedBuild = true;
+        return httpServerBuilder;
+    }
+
+    private TimeoutFromRequest grpcDetermineTimeout() {
+        return new TimeoutFromRequest() {
+            /**
+             * Return the timeout duration extracted from the GRPC timeout HTTP header if present or default timeout.
+             *
+             * @param request The HTTP request to be used as source of the GRPC timeout header
+             * @return The non-negative timeout duration which may be null
+             */
+            @Override
+            @Nullable
+            public Duration apply(HttpRequestMetaData request) {
+                Duration duration = DefaultGrpcClientBuilder.readTimeoutHeader(request);
+
+                duration = (null != duration) ?
+                        // As specified in the request
+                        duration
+                        // use server default (if any)
+                        : defaultTimeout;
+
+                if (null != duration) {
+                    try {
+                        AsyncContext.put(GRPC_DEADLINE_KEY, Instant.now().plus(duration));
+                    } catch (UnsupportedOperationException ignored) {
+                        // ignored -- async context has probably been disabled.
+                        // Timeout propagation will be partially disabled.
+                        // cancel()s will still happen which will accomplish the same effect though less efficiently
```

Comment:
Consider logging when this happens. It's good to notify users when they use a feature that depends on `AsyncContext`, but it's disabled. It may accidentelly happen if they pull a dependency that disables `AsyncContext`. 
Not sure what will be a good level for that, maybe `debug` is sufficient.